csharp/extractor/Semmle.Extraction.CSharp.Standalone/Razor.cs

```diff
@@ -0,0 +1,99 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using Semmle.Extraction.CSharp.Standalone;
+
+namespace Semmle.BuildAnalyser
+{
+    internal class Razor
+    {
+        private readonly DotnetVersion sdk;
+        private readonly ProgressMonitor progressMonitor;
+        private readonly DotNet dotNet;
+        private readonly string sourceGeneratorFolder;
+        private readonly string cscPath;
+
+        public Razor(DotnetVersion sdk, DotNet dotNet, ProgressMonitor progressMonitor)
+        {
+            this.sdk = sdk;
+            this.progressMonitor = progressMonitor;
+            this.dotNet = dotNet;
+
+            sourceGeneratorFolder = Path.Combine(this.sdk.FullPath, "Sdks/Microsoft.NET.Sdk.Razor/source-generators");
+            if (!Directory.Exists(sourceGeneratorFolder))
+            {
+                this.progressMonitor.RazorSourceGeneratorMissing(sourceGeneratorFolder);
+                throw new Exception($"Razor source generator folder {sourceGeneratorFolder} does not exist.");
+            }
+
+            cscPath = Path.Combine(this.sdk.FullPath, "Roslyn/bincore/csc.dll");
+            if (!File.Exists(cscPath))
+            {
+                this.progressMonitor.CscMissing(cscPath);
+                throw new Exception($"csc.dll {cscPath} does not exist.");
+            }
+        }
+
+        private static void GenerateAnalyzerConfig(string[] cshtmls, string analyzerConfigPath)
+        {
+            using var sw = new StreamWriter(analyzerConfigPath);
+            sw.WriteLine("is_global = true");
+
+            foreach (var f in cshtmls)
+            {
+                sw.WriteLine($"\n[{f}]");
+                var base64 = Convert.ToBase64String(Encoding.UTF8.GetBytes(f)); // TODO: this should be the relative path of the file.
+                sw.WriteLine($"build_metadata.AdditionalFiles.TargetPath = {base64}");
+            }
+        }
+
+        public IEnumerable<string> GenerateFiles(string[] cshtmls, string workingDirectory)
+        {
+            // TODO: the below command might be too long. It should be written to a temp file and passed to csc via @.
+
+            var name = Guid.NewGuid().ToString("N").ToUpper();
+            var analyzerConfig = Path.Combine(Path.GetTempPath(), name + ".txt");
+            var dllPath = Path.Combine(Path.GetTempPath(), name + ".dll");
+            var outputFolder = Path.Combine(workingDirectory, name);
+            Directory.CreateDirectory(outputFolder);
+
+            try
+            {
+                GenerateAnalyzerConfig(cshtmls, Path.Combine(sourceGeneratorFolder, analyzerConfig));
+
+                var args = new StringBuilder();
+                args.Append($"{cscPath} --do-not-trace /target:exe /generatedfilesout:{outputFolder} /out:{dllPath} /analyzerconfig:{analyzerConfig} ");
```

Comment:
`--do-not-trace` is an ugly workaround to the issue I faced with https://github.com/github/codeql/pull/13722/commits/78f3a3882774a5aa85af467851aa60909254b5d5. This call will only work if it is traced. The tracer will remove this argument. (When the standalone extractor is called directly (during debugging or `semmle-extractor-options: --standalone`), the call will fail.